### PR TITLE
NAS-119932 / 23.10 / Allow tying app version to SCALE releases

### PR DESCRIPTION
**Pull request by sonicaj**
This PR adds functionality to be able to specify app version to SCALE releases so for example if an app has specified that it can only be installed on Bluefin or above, Angelfish users should not be able to install or upgrade to that app version.
This essentially adds a baseline version which devs can specify so systems on that version or above that version can proceed with installation or upgrade of the app version in question.

### CODE REVIEW

**Review by Ornias1993 — COMMENTED** [review]
LGTM, this is extremely usefull :)

Altough you should really start providing docs for this in the `charts` repo in the docs folder there. Because you should provide devdocs for this so people can actually use it without being platform experts.

File: src/middlewared/middlewared/utils/__init__.py [hunk 1]
```diff
@@ -254,6 +255,11 @@ def sw_version():
     return VERSION
 
 
+@functools.cache
+def manifest_version():
```
Comment (yocalebo) [inline src/middlewared/middlewared/utils/__init__.py:259]: Please check the various system plugins. I believe we already have a method that returns the information you want. If there isn't, please rename this function to what it's actually doing. It's reading the manifest file to determine the scale release, not return the manifest file version?

**Review by yocalebo — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/utils/__init__.py (hunk 1 above)
Comment (sonicaj) [reply src/middlewared/middlewared/utils/__init__.py:259]: We don't unfortunately have that method already - we have a method which returns the verbose name or Full Name as we call it.
I'll rename the method to reflect reality.

Comment (yocalebo) [reply src/middlewared/middlewared/utils/__init__.py:259]: With https://github.com/truenas/middleware/commit/e42cd6e1da74b1bace5aa83a68db9b6dd8b02c31 merged, you can call `system.version False` so you don't have to add this helper function anymore.

File: src/middlewared/middlewared/plugins/catalogs_linux/items_util.py [hunk 2]
```diff
@@ -1,23 +1,59 @@
...
+) -> typing.Tuple[bool, bool]:
+    # `check_supported_key` is used because when catalog validation returns the data it only checks the
+    # missing features and based on that makes the decision. So if something is not already supported
+    # we do not want to validate minimum scale version in that case. However, when we want to report to
+    # the user as to why exactly the app version is not supported, we need to be able to make that distinction
+    if version_details['healthy'] and version_details['chart_metadata'].get('minimum_scale_version') and (
+        not check_supported_key or version_details['supported']
+    ):
+        try:
+            if sw_version(False) != version_details['chart_metadata']['minimum_scale_version'] and not can_update(
```
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/catalogs_linux/items_util.py:33]:
```suggestion
            if sw_version(fullname=False) != version_details['chart_metadata']['minimum_scale_version'] and not can_update(
```

File: src/middlewared/middlewared/plugins/update_/install_linux.py [hunk 3]
```diff
@@ -27,15 +28,12 @@ def install_manual_impl(self, job, path, dest_extracted, options=None):
...
-        with open(SCALE_MANIFEST_FILE) as f:
-            old_manifest = json.load(f)
-
         progress_callback(0, "Reading update file")
         with mount_update(path) as mounted:
             with open(os.path.join(mounted, "manifest.json")) as f:
                 manifest = json.load(f)
 
-            old_version = old_manifest["version"]
+            old_version = sw_version(False)
```
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/update_/install_linux.py:36]:
```suggestion
            old_version = sw_version(fullname=False)
```

**Review by yocalebo — CHANGES_REQUESTED** [review]

**Review by yocalebo — APPROVED** [review]

**Pull request merged**